src/coreComponents/physicsSolvers/multiphysics/poromechanicsKernels/ThermalMultiphasePoromechanics_impl.hpp

```diff
@@ -0,0 +1,664 @@
+        edgeManager,
+        faceManager,
+        targetRegionIndex,
+        elementSubRegion,
+        finiteElementSpace,
+        inputConstitutiveType,
+        inputDispDofNumber,
+        rankOffset,
+        inputMatrix,
+        inputRhs,
+        gravityVector,
+        inputFlowDofKey,
+        numComponents,
+        numPhases,
+        fluidModelKey ),
+  m_rockInternalEnergy_n( inputConstitutiveType.getInternalEnergy_n() ),
+  m_rockInternalEnergy( inputConstitutiveType.getInternalEnergy() ),
+  m_dRockInternalEnergy_dTemperature( inputConstitutiveType.getDinternalEnergy_dTemperature() ),
+  m_temperature_n( elementSubRegion.template getField< fields::flow::temperature_n >() ),
+  m_temperature( elementSubRegion.template getField< fields::flow::temperature >() )
+{
+  // extract fluid constitutive data views
+  {
+    string const fluidModelName = elementSubRegion.template getReference< string >( fluidModelKey );
+    constitutive::MultiFluidBase const & fluid =
+      elementSubRegion.template getConstitutiveModel< constitutive::MultiFluidBase >( fluidModelName );
+
+    m_fluidPhaseInternalEnergy_n = fluid.phaseInternalEnergy_n();
+    m_fluidPhaseInternalEnergy = fluid.phaseInternalEnergy();
+    m_dFluidPhaseInternalEnergy = fluid.dPhaseInternalEnergy();
+  }
+}
+
+
+template< typename SUBREGION_TYPE,
+          typename CONSTITUTIVE_TYPE,
+          typename FE_TYPE >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void ThermalMultiphasePoromechanics< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >::
+setup( localIndex const k,
+       StackVariables & stack ) const
+{
+  // initialize displacement, pressure, temperature and fluid components degree of freedom
+  Base::setup( k, stack );
+
+  stack.localTemperatureDofIndex = stack.localPressureDofIndex + m_numComponents + 1;
+  stack.temperature = m_temperature[k];
+  stack.deltaTemperatureFromLastStep = m_temperature[k] - m_temperature_n[k];
+}
+
+template< typename SUBREGION_TYPE,
+          typename CONSTITUTIVE_TYPE,
+          typename FE_TYPE >
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void ThermalMultiphasePoromechanics< SUBREGION_TYPE, CONSTITUTIVE_TYPE, FE_TYPE >::
+smallStrainUpdate( localIndex const k,
+                   localIndex const q,
+                   StackVariables & stack ) const
+{
+  real64 porosity = 0.0;
+  real64 porosity_n = 0.0;
+  real64 dPorosity_dVolStrain = 0.0;
+  real64 dPorosity_dPressure = 0.0;
+  real64 dPorosity_dTemperature = 0.0;
+  real64 dSolidDensity_dPressure = 0.0;
+
+  // Step 1: call the constitutive model to update the total stress, the porosity and their derivatives
+  m_constitutiveUpdate.smallStrainUpdatePoromechanics( k, q,
+                                                       m_pressure_n[k],
+                                                       m_pressure[k],
+                                                       stack.temperature,
+                                                       stack.deltaTemperatureFromLastStep,
```

Comment:
since we pass `pressure` and `pressure_n` we could to the same for the `temperature` just to be consistent.

Reply:
Yes, you are right. Once this PR is merged, I will immediately do another one to:
- Change the way we pass temperature here
- Clean up a few things in the `BiotPorosity` model
- Add an integrated test for sequential multiphase THM, since I am adding the pieces of code that you pointed me to to activate the sequential coupling in the present PR.